File: algorithms/quicksort/quicksort.go
package quicksort

import "fmt"

func Quicksort(A []int, p, r int) {
	if p < r {
		i := Partition(A, p, r)
		fmt.Println(i)
		Quicksort(A, p, i-1)
		Quicksort(A, i+1, r)
	}
}

func Partition(A []int, p, r int) int {
	x := A[r]
	i := p - 1

	for j := p; j < r; j++ {
		if A[j] <= x {
			i += 1
			fmt.Println(A)
			A[i], A[j] = A[j], A[i]
			fmt.Println(A)
			fmt.Println()
		}
	}

	A[i+1], A[r] = A[r], A[i+1]

	return i + 1
}

File: algorithms/quicksort/quicksort.c
#include <stdlib.h>
#include <stdio.h>

int partition(int* A,int lo, int hi);

void quicksort(int* A, int lo, int hi) {
	if(lo < hi) {
		int p = partition(A, lo, hi);
		quicksort(A, lo, p - 1);
		quicksort(A, p + 1, hi);
	}
}

int partition(int* A, int lo, int hi) {
	int pivot = A[hi];
	int i = lo - 1;
	for(int j = lo; j < hi - 1; j++) {
		if(A[j] < pivot) {
			i = i + 1;
			int tmp = A[i];;
			A[i] = A[j];
			A[j] = tmp;
		}

		int t = A[i+1];
		A[i+1] = hi;
		A[hi] = t;
	}

	return i + 1;
}

int main(int argc, char** argv) {
	int a[] = {1,2,3,3,5,5};
	quicksort(a, 0, (sizeof(a) / sizeof(int)) - 1);

	for(int i = 0; i < sizeof(a) / sizeof(int); i++) {
		printf("%d, ", a[i]);
	}

	return 0;
}
